### ISSUE TITLE
Worker hangs if 'Gossip' is loaded after 'Heart' bootstep.

### ISSUE THREAD

**Issue by arhoyling** [celery/celery#4257]
## Checklist

- [X] I have included the output of ``celery -A proj report`` in the issue.
      (if you are not able to do this, then at least specify the Celery
       version affected).
   ```
   software -> celery:3.1.25 (Cipater) kombu:3.0.37 py:2.7.13
               billiard:3.3.0.23 librabbitmq:1.6.1
   platform -> system:Linux arch:64bit, ELF imp:CPython
   loader   -> celery.loaders.app.AppLoader
   settings -> transport:librabbitmq results:disabled
   ```
   Worker info
   ```
    -------------- XXX@XXX v3.1.25 (Cipater)
   ---- **** ----- 
   --- * ***  * -- Linux-4.4.0-93-generic-x86_64-with-debian-stretch-sid
   -- * - **** --- 
   - ** ---------- [config]
   - ** ---------- .> app:         XXX:0x7f4a6dff05d0
   - ** ---------- .> transport:   amqp://XXX
   - ** ---------- .> results:     disabled://
   - *** --- * --- .> concurrency: {min=0, max=3} (prefork)
   -- ******* ---- 
   --- ***** ----- [queues]
    -------------- .> default     exchange=default(direct) key=default
   ```

- [ ] I have verified that the issue exists against the `master` branch of Celery.

## Steps to reproduce
Repeat until `Heart` precedes `Gossip` in boot order*:
 1. Start a single worker.
 1. Issue a SIGTERM signal, a stop command, or Ctrl-Break to the worker process.
 1. Ping the worker.

\* This ought to happen as there is no dependency between the two and `Blueprint` uses a standard `dict` in its `claim_steps()` method (so the order is non-deterministic for Python < 3.6).
## Expected behavior
The worker should stop when issued a warm shutdown as it does not have any running task yet.

## Actual behavior
The process hangs forever until the `ping` command is issued at which point it proceeds to shutting down.

`strace` logs after sending SIGTERM to the worker:

```
sendto(13, "\1\0\1\0\0\0\v\0<\0\n\0\0\0\0\0\1\1\316", 19, MSG_NOSIGNAL, NULL, 0) = 19
recvfrom(13, "\1\0\1\0\0\0\4\0<\0\v\316", 131072, 0, NULL, NULL) = 12
recvfrom(13, 0x28901f0, 131072, 0, NULL, NULL) = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGTERM {si_signo=SIGTERM, si_code=SI_USER, si_pid=27645, si_uid=1762} ---
rt_sigreturn({mask=[]})        = -1 EINTR (Interrupted system call)
recvfrom(13, 
```

The consumer is expecting a message which is never coming.

**Comment by arhoyling** [comment]
The consumer stalls at the first call to amqp's `drain_events()`.

Seems related to #1847 (Worker hangs at startup when running without mingle and gossip and many messages in queue).

**Comment by arhoyling** [comment]
Updating amqp to the latest version 2.2.1 seems to have solved the problem.

(Probably fixed by commit [`9fccfb7`](https://github.com/celery/py-amqp/pull/144/commits/9fccfb762486ae1b8e42d44560af7b9c2480d8b3))
